packages/runtime/container-runtime/src/test/containerRuntime.spec.ts

```diff
@@ -1465,5 +1513,172 @@ describe("Runtime", () => {
+				await yieldEventLoop();
+
+				const summarizeResultP = containerRuntime.submitSummary({
+					summaryLogger: createChildLogger(),
+					cancellationToken: neverCancelledSummaryToken,
+				});
+
+				// Advance the clock by the time that container runtime would wait for pending ops to be processed.
+				clock.tick(defaultPendingOpsWaitTimeoutMs);
+				const summarizeResult = await summarizeResultP;
+				assert(summarizeResult.stage === "base", "Summary did not fail");
+				assert.strictEqual(
+					summarizeResult.error.message,
+					"PendingOpsWhileSummarizing",
+					"Summary did not fail with the right error",
+				);
+				assert.strictEqual(
+					summarizeResult.error.beforeGenerate,
+					true,
+					"It should have failed before generating summary",
+				);
+			});
+
+			it("summary fails after generate if there are pending ops", async () => {
+				// Patch the summarize function to submit messages during it. This way there will be pending
+				// messages after generating the summary.
+				const patch = (fn: (...args) => Promise<ISummaryTreeWithStats>) => {
+					const boundFn = fn.bind(containerRuntime);
+					return async (...args: any[]) => {
+						// Submit an op and yield for it to be flushed from outbox to pending state manager.
+						containerRuntime.submitDataStoreOp("fakeId", "fakeContents");
+						await yieldEventLoop();
+						return boundFn(...args);
+					};
+				};
+				containerRuntime.summarize = patch(containerRuntime.summarize);
+
+				const summarizeResult = await containerRuntime.submitSummary({
+					summaryLogger: createChildLogger(),
+					cancellationToken: neverCancelledSummaryToken,
+				});
+				assert(summarizeResult.stage === "base", "Summary did not fail");
+				assert.strictEqual(
+					summarizeResult.error.message,
+					"PendingOpsWhileSummarizing",
+					"Summary did not fail with the right error",
+				);
+				assert.strictEqual(
+					summarizeResult.error.beforeGenerate,
+					false,
+					"It should have failed after generating summary",
+				);
+			});
+
+			it("summary passes if pending ops are processed during pending op processing timeout", async () => {
+				// Create a container runtime type where the submit method is public. This makes it easier to test
+				// submission and processing of ops. The other option is to send data store or alias ops whose
+				// processing requires creation of data store context and runtime as well.
+				type ContainerRuntimeWithSubmit = Omit<ContainerRuntime, "submit"> & {
+					submit(
+						containerRuntimeMessage: ContainerRuntimeMessage,
+						localOpMetadata: unknown,
+						metadata: Record<string, unknown> | undefined,
+					): void;
+				};
+				const containerRuntimeWithSubmit =
+					containerRuntime as unknown as ContainerRuntimeWithSubmit;
+				// Submit a rejoin op and yield for it to be flushed from outbox to pending state manager.
+				containerRuntimeWithSubmit.submit(
+					{
+						type: ContainerMessageType.Rejoin,
```

Comment:
Idea - You could use compatDetails and a bogus type here too, to ensure no side effects.

Reply:
Actually don't do that, it wouldn't make sense for a local op.  Rejoin is probably tame enough anyway

Reply:
The only problem is that rejoin may be removed. But then compiler would complain and the test would be fixed.